feat(inbound): Log traffic with the 'audit' policy

alpeb opened the pull request:
Audit mode is triggered by the policy controller, which will create an authorization named "audit" allowing traffic for the given target. When the proxy processes an authorization with such name it will log it at INFO.

Also, add "audit" to the possible values for
`LINKERD2_PROXY_INBOUND_DEFAULT_POLICY`, whose effect is the same as "all-unauthenticated".

Please check linkerd/website#1805 for how this is supposed to work from the user's perspective.

mateiidavid commented on linkerd/app/src/env.rs:
```diff
@@ -1008,7 +1008,7 @@ fn parse_default_policy(
             Ok(inbound::policy::defaults::all_authenticated(detect_timeout).into())
         }
-        "all-unauthenticated" => {
+        "all-unauthenticated" | "audit" => {
```
question: if we treat this as an `all-unauthenticated` by default, when we create the default policy will it still have the correct metadata?

What would happen if for example we have an `audit` default policy and we get an inbound connection? Will we still log with the audit labels?

mateiidavid reviewed (commented):
Left a small question on parsing default policies. I was also wondering if it's worth adding some tests.

Integration tests have a policy controller mock that implements the inbound and outbound interfaces (see https://github.com/linkerd/linkerd2-proxy/blob/main/linkerd/app/integration/src/policy.rs). There are also some unit tests in:
* Inbound TCP stack (https://github.com/linkerd/linkerd2-proxy/blob/main/linkerd/app/inbound/src/accept.rs#L154-L170)
* Inbound policy module (https://github.com/linkerd/linkerd2-proxy/blob/main/linkerd/app/inbound/src/policy.rs#L211).

I wonder if it makes sense to add a test or two to ensure this all behaves as expected.

alpeb commented:
I've added a new new branch in the matcher parsing LINKERD2_PROXY_INBOUND_DEFAULT_POLICY that creates a similiar ServerPolicy as the one used for all_unauthenticated, but named "audit", this name being the only thing that causes the special log entry to be produced. I     tested this manually using a proxy with a hard-coded `policy::Config::Fixed` policy with an audit default ServerPolicy to avoid overriding LINKERD2_PROXY_INBOUND_DEFAULT_POLICY's policy with something coming from the policy controller.

As for unit/integration tests, I couldn't quite find a place for this that didn't duplicate what we already have for all_unauthenticated. We could add something with some tracing subscriber that would catch the log entry but not sure if it's worth it.     Open to more suggestions.

olix0r commented on linkerd/app/inbound/src/policy/http.rs:
```diff
@@ -202,7 +202,25 @@ impl<T, N> HttpPolicyService<T, N> {
         {
-            Some(authz) => authz,
+            Some(authz) => {
+                if authz.meta.name() == "audit" {
```
in all of these cases don't we want to check that we're dealing with a kind=default and not just an arbitrary authorization resource named audit?

alpeb replied:
Good point, should be fixed now.

olix0r commented on linkerd/proxy/server-policy/src/meta.rs:
```diff
@@ -39,6 +39,10 @@ impl Meta {
     }
+
+    pub fn is_audit(&self) -> bool {
+        self.kind() == "default" && self.name() == "audit"
```
We can use pattern matching to avoid string comparison here:

```suggestion
        matches!(self, Self::Default { name } if name == "audit")
```
